### ISSUE TITLE
UI needs to respect device settings for height and width

### ISSUE THREAD

**Issue by r0bc94** [sbs20/scanservjs#45]
Hey, 

I have the following problem while trying to scan an image using scanservjs and sane. 

When ever I try to obtain a preview, scanservjs tries to execute the following command: 
```bash
scanimage -d "hpaio:/usb/Deskjet_F2400_series?serial=CN03GB749405H4" --mode "Color" --resolution 100 -l 0 -t 0 -x 215 -y 297 --formattiff --brightness 0 --contrast 0 > "./data/preview/preview.tif"
```

The problem is, that the command throws the following error: 
```
scanimage: unrecognized option '--formattiff'
```

Also, when I try to directly create a scan, the following error is thrown: 
```
scanimage: setting of option --br-y failed (Invalid argument)
```

Is there anything wrong here? Keep in mind that I've installed sane locally on my raspberry pi. Should I rather use the provided docker package? 

I hope, you can help me with this issue.

**Comment by JnCrMx** [comment]
That seems to be a problem with the command lines arguments passed to ``scanimage`` and not with sane itself.
A locally installed sane on a Raspberry Pi should work fine.

Those errors seem really weird to me to be honest because it seems like some spaces are missing (e.g. there should be a space between ``--format`` and ``tiff``).

Could you look into the logs (use ``journalctl -u scanservjs`` if you are running it using the systemd service) and see if there are any odd messages/errors? It would be nice and useful if you could attach some logs.

**Comment by r0bc94** [comment]
This is actually my bad. After I saw the errors coming from `scanimage`, I searche'd the logs for the executed command. It seems that I've accidentally removed the space between `--format` and `tiff`. 

Trusting the logs, the space is inserted correctly: 
```
System.execute:start=/usr/bin/scanimage -d "hpaio:/usb/Deskjet_F2400_series?serial=CN03GB749405H4" --mode "Color" --resolution 100 -l 0 -t 0 -x 215 -y 297 --format tiff --brightness 0 --contrast 0 > "./data/preview/preview.tif"
```

However, its still not possible to create a preview or a scan since the following error is thrown by the backend: 
```
Command failed: /usr/bin/convert  ./data/preview/preview.tif ./data/preview/preview.jpg↵convert: Cannot read TIFF header. `./data/preview/preview.tif' @ error/tiff.c/TIFFErrors/564.↵convert: no images defined `./data/preview/preview.jpg' @ error/convert.c/ConvertImageCommand/3258.↵
```

**Comment by JnCrMx** [comment]
Hmmm...

The command line for calling ``scanimage`` looks good to me.

Could you try to execute ``/usr/bin/scanimage -d "hpaio:/usb/Deskjet_F2400_series?serial=CN03GB749405H4" --mode "Color" --resolution 100 -l 0 -t 0 -x 215 -y 297 --format tiff --brightness 0 --contrast 0 > preview.tif`` manually and check if it works?

If not, there seems to be something wrong about sane as you suspected before, so please attach the output of ``scanimage --version`` in that case (which might be helpful to reproduce the error).

You could also try to remove some of the options (e.g. ``-l 0 -t 0 -x 215 -y 297``) to see if that changes anything.

**Comment by r0bc94** [comment]
When I try to execute the command from your comment above, I get the same error: 
```
$ sudo /usr/bin/scanimage -d "hpaio:/usb/Deskjet_F2400_series?serial=CN03GB749405H4" --mode "Color" --resolution 100 -l 0 -t 0 -x 215 -y 297 --format tiff --brightness 0 --contrast 0 > preview.tif
scanimage: setting of option --br-y failed (Invalid argument)
```

Strangely, if I only omit the `-y`, but keep the `-x` parameter, everything works fine. 

The installed version of sane and scanimage is: 
```
$ scanimage --version
scanimage (sane-backends) 1.0.25; backend version 1.0.25
```

(Its porbably a very old version of scanimage and sane, since I'm still on raspbian stretch. I plan to update in the near future).

**Comment by JnCrMx** [comment]
That behavior is indeed really strange, but seems to be indeed related to an old version of ``sane`` and ``scanimage``.

Using version ``1.0.27`` I didn't experience such issues. You should really consider upgrading to ``buster``.
I tried to reproduce that issue in a ``chroot`` (because I don't want to downgrade my Pi to ``stretch``), but ``scanimage`` doesn't seem to work in it.

There might be a few "quicker" solutions for you (if you don't want to upgrade now):

- I wrote a little [shell script](https://github.com/sbs20/scanservjs/files/4899496/scanimage_wrapper.sh.txt) that removes the ``-y`` parameter from the arguments and then calls ``scanimage``.  Using that script as a wrapper for ``scanimage`` might also solve your problem temporarily.
- You could also try to compile ``sane`` and ``scanimage`` from source (found at https://gitlab.com/sane-project/backends/) and either replace your current version (kinda dangerous) or use only the ``scanimage`` binary (which might not work very well, but might be worth a try).

For both solutions, it might be useful to replace the path for scanimage in ``/var/www/scanservjs/classes/Config.js``.
Simply modify the line saying ``Scanimage: '/usr/bin/scanimage',``.

I hope this isn't sounding too complex 😅

**Comment by sbs20** [comment]
As @JnCrMx  says, this seems to be an issue with `sane` - looking at the various [search results](https://www.google.com/search?q=setting+of+option+--br-y+failed+(Invalid+argument)) it could be that the `y` value is out of range. Could you try a command line with a lower y value e.g. 50? It won't fix anything but might give us a clue as to what's going on?

It might also be useful to see what the scanner reports its possible values are. Could you post the `device.conf` here? Or just the snippet from `cat device.conf | grep '\-y' -A 3`?

**Comment by r0bc94** [comment]
Thanks for your responses :). 

As @sbs20 suggested, I tried to call `scanimage` with a lower value for `-y` and it worked. Also, the y - range for the scanner inside the `devices.conf` file is set to `0..296.926`. See: 

```
 $ cat device.conf | grep '\-y' -A 3
        "-y": {
            "default": "296.926",
            "options": "0..296.926mm"
        }
```

For whatever reason, the web api defaults to a y - value of 297, as you can see in the `scanimage` command, which is executed by the backend. Maybe this is some sort of rounding error inside the front- or backend? (I didn't look at the configuration which is send from the frontend to the backend when starting a new scan yet). 

Either way, it seems to work just fine, if I lower the y - value to 296.

**Comment by JnCrMx** [comment]
The preview function appears to be using ``Config.MaximumScanHeightInMm`` as a default value for ``-y`` regardless of the supported range. (refer to https://github.com/sbs20/scanservjs/blob/master/classes/Api.js#L67, https://github.com/sbs20/scanservjs/blob/master/classes/ScanRequest.js#L65 and https://github.com/sbs20/scanservjs/blob/master/classes/Config.js#L13).

Please try changing that to ``296`` (you can find it in ``/var/www/scanservjs/classes/Config.js``).

@sbs20, do you think it would be good to parse that value from ``scanimage -A`` (e.g. just use the identified maximum bounds for preview scans)?

**Comment by r0bc94** [comment]
@JnCrMx This actually worked! Thank you :smile: 

IMO, a welcome addition would be, that the default values of the `width` and `height` fields in the web frontend should default to the maximum width and height values, which are specified in the server's config file. 

Anyway, from my point of view, this issue can be closed.

**Comment by sbs20** [comment]
@JnCrMx Yes - I think the work in the current development branch needs to continue, then we read the maximum values for preview from the result, as well as setting the UI values from there. That would resolve the issue more satisfactorily.

@r0bc94 Thanks for raising the issue. We'll keep it open to track the status here.

**Comment by JnCrMx** [comment]
About the ``development`` branch:

I've pushed a few commits (in my fork) to improve the ``decorate`` function. It is now able to parse ``(in steps of %d)`` as well as units (``mm``, ``dpi``, ``%``, etc.).

It would be nice if you could have a short look at it. I've tested it with both ``device.conf`` and modified them a bit to see if it also works with more complex configurations, but there might be some aspects I missed.

If that function does fine, we can start working on modifying the UI and the preview function.

**Comment by sbs20** [comment]
This is partially resolved in v2.0.2 but needs a bit of further work to sort the client side aspect ratio out.
